# 12 - Final Test

**This is the final test. `test.parquet` is loaded ONCE, in this notebook, for the
first and only time in the whole project.**

The model configuration is FROZEN - hyperparameters, thresholds and features were
fixed in notebook 11, looking only at validation (2014). Nothing here is re-tuned based
on what this notebook shows. If the result is disappointing, it is reported as
it is.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, average_precision_score, brier_score_loss
from xgboost import XGBClassifier
import warnings

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', None)

%matplotlib inline

PROCESSED_DIR = Path('..') / 'data' / 'processed'

train = pd.read_parquet(PROCESSED_DIR / 'train.parquet')
validation = pd.read_parquet(PROCESSED_DIR / 'validation.parquet')
test = pd.read_parquet(PROCESSED_DIR / 'test.parquet')
transfer_60m = pd.read_parquet(PROCESSED_DIR / 'transfer_60m.parquet')

print(f'train: {train.shape} | validation: {validation.shape} | test: {test.shape} | transfer_60m: {transfer_60m.shape}')


train: (172988, 89) | validation: (162570, 89) | test: (282787, 89) | transfer_60m: (54969, 89)


### Rebuilding FEATURE_SET, financials, prepare_X (identical to notebooks 06-11)

In [2]:
EVAL_ONLY = ['loan_status', 'loan_amnt', 'installment', 'term', 'total_rec_prncp']
PROVISIONAL_EXCLUDE = ['int_rate', 'grade', 'sub_grade']

family_C_features = ['funded_amnt', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
    'purpose', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'fico_range_low', 'fico_range_high',
    'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec',
    'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'collections_12_mths_ex_med',
    'mths_since_last_major_derog', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
    'tot_cur_bal', 'total_rev_hi_lim', 'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy',
    'bc_util', 'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
    'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc',
    'mths_since_recent_bc', 'mths_since_recent_bc_dlq', 'mths_since_recent_inq',
    'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl',
    'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 'num_rev_accts',
    'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m', 'num_tl_30dpd', 'num_tl_90g_dpd_24m',
    'num_tl_op_past_12m', 'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pub_rec_bankruptcies',
    'tax_liens', 'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit',
    'total_il_high_credit_limit', 'emp_length_anos']
assert len(family_C_features) == 65

engineered_flags = ['era_pre_2012',
                     'mths_since_last_delinq_missing', 'mths_since_last_record_missing',
                     'mths_since_recent_bc_dlq_missing', 'mths_since_recent_revol_delinq_missing',
                     'mths_since_last_major_derog_missing', 'emp_length_missing',
                     'mths_since_recent_inq_missing', 'num_tl_120dpd_2m_missing', 'sparse_bureau_missing']
assert len(engineered_flags) == 10

new_features = ['installment_to_income', 'loan_to_income', 'credit_history_months',
                 'revol_bal_to_income', 'open_acc_ratio']
assert len(new_features) == 5

redundant_cols = {'fico_range_high': 'redundant (r=1.0 with fico_range_low)'}
FEATURE_SET = [c for c in family_C_features if c not in redundant_cols] + engineered_flags + new_features
assert len(FEATURE_SET) == 79

CATEGORICAL_COLS = ['home_ownership', 'purpose', 'verification_status', 'initial_list_status', 'application_type']
REFERENCE_DATE = pd.Timestamp('2000-01-01')


def compute_financials(df):
    interest = (df['installment'] * df['term']) - df['loan_amnt']
    loss_raw = df['loan_amnt'] - df['total_rec_prncp']
    return interest, loss_raw.clip(lower=0)


def prepare_X(df, feature_cols, categorical_cols):
    X = df[feature_cols].copy()
    for c in ['issue_d', 'earliest_cr_line']:
        if c in X.columns:
            X[c] = (X[c] - REFERENCE_DATE).dt.days
    cat_present = [c for c in categorical_cols if c in X.columns]
    X = pd.get_dummies(X, columns=cat_present, drop_first=True)
    return X


def profit_at_threshold(y_true, y_prob, threshold, interest, loss):
    y_true = np.asarray(y_true)
    y_prob = np.asarray(y_prob)
    interest = np.asarray(interest)
    loss = np.asarray(loss)
    approved = y_prob < threshold
    return interest[approved & (y_true == 0)].sum() - loss[approved & (y_true == 1)].sum()


def decompose(y_prob, threshold, y_true, interest, loss, loan_amnt):
    y_prob = np.asarray(y_prob)
    y_true = np.asarray(y_true)
    interest = np.asarray(interest)
    loss = np.asarray(loss)
    loan_amnt = np.asarray(loan_amnt)
    rejected = y_prob >= threshold
    rejected_co = rejected & (y_true == 1)
    rejected_fp = rejected & (y_true == 0)
    avoided_loss = loss[rejected_co].sum()
    lost_interest = interest[rejected_fp].sum()
    return {'threshold': threshold, 'n_rejected': int(rejected.sum()),
            'total_rejected_amount': loan_amnt[rejected].sum(),
            'n_rejected_charged_off': int(rejected_co.sum()), 'n_rejected_fully_paid': int(rejected_fp.sum()),
            'avoided_loss_$': avoided_loss, 'lost_interest_$': lost_interest,
            'net_gain_$': avoided_loss - lost_interest}


threshold_grid = np.round(np.arange(0.01, 1.0, 0.01), 2)

THRESH_M1 = 0.38
THRESH_XGB = 0.31
XGB_WF_PARAMS = dict(max_depth=8, learning_rate=0.03, n_estimators=600, min_child_weight=10,
                     subsample=0.8, colsample_bytree=0.6)

print('Infrastructure rebuilt. Frozen thresholds: M1 =', THRESH_M1, '| XGB_walkforward =', THRESH_XGB)


Infrastructure rebuilt. Frozen thresholds: M1 = 0.38 | XGB_walkforward = 0.31


## Preprocessing rule - check before predicting

Everything fitted on the training set is applied to the test set without refitting. Before generating any
prediction, we confirm that the columns of `X_test` are IDENTICAL to those of `X_train` - same
set, same order. If they are not, the notebook stops here.

In [3]:
X_train = prepare_X(train, FEATURE_SET, CATEGORICAL_COLS)
X_test = prepare_X(test, FEATURE_SET, CATEGORICAL_COLS)

cols_train = X_train.columns.tolist()
cols_test = X_test.columns.tolist()

if cols_train != cols_test:
    only_train = set(cols_train) - set(cols_test)
    only_test = set(cols_test) - set(cols_train)
    print('COLUMN MISMATCH BETWEEN TRAIN AND TEST.')
    print('Columns only in train:', only_train)
    print('Columns only in test:', only_test)
    print('Same order (if same set)?', cols_train == cols_test)
    raise RuntimeError('Train and test columns differ - preprocessing stopped as a safeguard.')
else:
    print(f'OK: {len(cols_train)} identical columns, same order, between train and test.')


OK: 90 identical columns, same order, between train and test.


## Section 0 - Test set sanity checks (no performance yet)

In [4]:
y_test = test['target'].values
n_test = len(test)
default_rate_test = test['target'].mean()

print(f'Test N: {n_test}')
print(f'Default rate in test: {default_rate_test * 100:.4f}%')
print(f'issue_d range in test: {test["issue_d"].min()} to {test["issue_d"].max()}')
print(f'Is the whole test set 2015? {(test["issue_d"].dt.year == 2015).all()}')


Test N: 282787
Default rate in test: 14.8836%
issue_d range in test: 2015-01-01 00:00:00 to 2015-12-01 00:00:00
Is the whole test set 2015? True


**About the index-intersection check:** the parquet files were saved with the
index reset (0..N-1) independently in each split - the raw DataFrame index is NOT
a global identifier (index 0 of test has nothing to do with index 0
of train). The equivalent leakage check is by row CONTENT: no test
row may be identical, column by column, to any train row.

In [5]:
train_hashes = set(pd.util.hash_pandas_object(train, index=False))
test_hashes = set(pd.util.hash_pandas_object(test, index=False))
overlap = train_hashes & test_hashes

print(f'Test rows whose content is identical to some train row: {len(overlap)}')
assert len(overlap) == 0, 'WARNING: there are duplicate rows between train and test.'
print('OK: no test row is identical to a train row.')


Test rows whose content is identical to some train row: 0
OK: no test row is identical to a train row.


**SMD (standardized mean difference) between train and test, per feature - the 10
largest.** It shows how much the 2015 population differs from the training population
(2007-2013); some degree of shift was already expected (seen on the 2014 validation).

In [6]:
def smd(a, b):
    mean_diff = a.mean() - b.mean()
    pooled_std = np.sqrt((a.var(ddof=1) + b.var(ddof=1)) / 2)
    return mean_diff / pooled_std if pooled_std > 0 else 0.0

smd_rows = []
for col in X_train.columns:
    smd_rows.append({'feature': col, 'smd': smd(X_train[col], X_test[col])})
smd_df = pd.DataFrame(smd_rows)
smd_df['abs_smd'] = smd_df['smd'].abs()
smd_df = smd_df.sort_values('abs_smd', ascending=False).drop(columns='abs_smd').reset_index(drop=True)
smd_df.head(10)


,feature,smd
0,issue_d,-3.251076
1,initial_list_status_w,-0.927681
2,era_pre_2012,0.922628
3,num_tl_30dpd,-0.916133
4,pct_tl_nvr_dlq,-0.844014
5,mo_sin_old_il_acct,0.808631
6,num_tl_120dpd_2m_missing,0.695062
7,num_tl_120dpd_2m,-0.694251
8,percent_bc_gt_75,0.689111
9,bc_util,0.687600


## Section 1 - Train the two finalists on the training set (frozen configuration)

Hyperparameters and thresholds are exactly the ones fixed in notebook 11. No tuning
happens in this notebook. XGBoost training uses the original row order of
`train.parquet` (not re-sorted by `issue_d`), consistent with the root-cause correction
recorded in notebook 11.

In [7]:
y_train = train['target'].values

scaler = StandardScaler()
X_train_scaled = pd.DataFrame(scaler.fit_transform(X_train), columns=X_train.columns, index=X_train.index)
X_test_scaled = pd.DataFrame(scaler.transform(X_test), columns=X_test.columns, index=X_test.index)

with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter('always')
    m1_model = LogisticRegression(penalty='l2', solver='lbfgs', max_iter=2000, random_state=42)
    m1_model.fit(X_train_scaled, y_train)
    convergence_warnings = [x for x in w if 'onverg' in str(x.message)]

n_iter = m1_model.n_iter_[0]
print(f'M1: n_iter_={n_iter} (max_iter=2000) | ConvergenceWarning raised? {len(convergence_warnings) > 0}')
if n_iter >= 2000 or len(convergence_warnings) > 0:
    raise RuntimeError('M1 did not converge within max_iter=2000.')
print('M1 converged.')

xgb_wf = XGBClassifier(**XGB_WF_PARAMS, random_state=42, eval_metric='logloss', n_jobs=1)
xgb_wf.fit(X_train, y_train)
print('XGB_walkforward trained.')


M1: n_iter_=84 (max_iter=2000) | ConvergenceWarning raised? False
M1 converged.


XGB_walkforward trained.


In [8]:
# Predictions on test (evaluation) and on validation (only to recompute the numbers already
# fixed in notebook 11, for the Section 4 comparison - no tuning is based on them)
X_val = prepare_X(validation, FEATURE_SET, CATEGORICAL_COLS)
X_val = X_val.reindex(columns=X_train.columns, fill_value=0)
X_val_scaled = pd.DataFrame(scaler.transform(X_val), columns=X_val.columns, index=X_val.index)
y_val = validation['target'].values

interest_test, loss_test = compute_financials(test)
interest_test = interest_test.values
loss_test = loss_test.values
loan_amnt_test = test['loan_amnt'].values

interest_val, loss_val = compute_financials(validation)
interest_val = interest_val.values
loss_val = loss_val.values

y_prob_test_m1 = m1_model.predict_proba(X_test_scaled)[:, 1]
y_prob_test_xgb = xgb_wf.predict_proba(X_test)[:, 1]

y_prob_val_m1 = m1_model.predict_proba(X_val_scaled)[:, 1]
y_prob_val_xgb = xgb_wf.predict_proba(X_val)[:, 1]

print('Predictions generated on test and validation (validation for reference only, Section 4).')


Predictions generated on test and validation (validation for reference only, Section 4).


### Bootstrap on the test set (1,000 resamples, seed=42) - feeds Sections 2 and 3

A single bootstrap over the test set, storing per resample: AUC-ROC, AUC-PR and Brier of
M1/XGB, and the profit of M0b/M1/XGB at the frozen threshold. Sections 2 and 3 are read from
this same set of resamples.

In [9]:
N_BOOT = 1000
SEED = 42
rng = np.random.default_rng(SEED)

boot = {
    'auc_m1': np.zeros(N_BOOT), 'auc_xgb': np.zeros(N_BOOT),
    'auc_pr_m1': np.zeros(N_BOOT), 'auc_pr_xgb': np.zeros(N_BOOT),
    'brier_m1': np.zeros(N_BOOT), 'brier_xgb': np.zeros(N_BOOT),
    'profit_m0b': np.zeros(N_BOOT), 'profit_m1': np.zeros(N_BOOT), 'profit_xgb': np.zeros(N_BOOT),
}

for b in range(N_BOOT):
    idx = rng.integers(0, n_test, size=n_test)
    yb = y_test[idx]
    ib = interest_test[idx]
    lb = loss_test[idx]
    pm1 = y_prob_test_m1[idx]
    pxgb = y_prob_test_xgb[idx]

    boot['auc_m1'][b] = roc_auc_score(yb, pm1)
    boot['auc_xgb'][b] = roc_auc_score(yb, pxgb)
    boot['auc_pr_m1'][b] = average_precision_score(yb, pm1)
    boot['auc_pr_xgb'][b] = average_precision_score(yb, pxgb)
    boot['brier_m1'][b] = brier_score_loss(yb, pm1)
    boot['brier_xgb'][b] = brier_score_loss(yb, pxgb)
    boot['profit_m0b'][b] = ib[yb == 0].sum() - lb[yb == 1].sum()
    boot['profit_m1'][b] = profit_at_threshold(yb, pm1, THRESH_M1, ib, lb)
    boot['profit_xgb'][b] = profit_at_threshold(yb, pxgb, THRESH_XGB, ib, lb)

print(f'Bootstrap done. N_BOOT={N_BOOT} seed={SEED}')


Bootstrap done. N_BOOT=1000 seed=42


## Section 2 - Test results

Point estimate + 95% CI (percentile, from the 1,000 resamples above) - not point values with
false precision.

In [10]:
def ci(arr):
    lo, hi = np.percentile(arr, [2.5, 97.5])
    return lo, hi

profit_m0b_test = interest_test[y_test == 0].sum() - loss_test[y_test == 1].sum()
approved_m1_test = y_prob_test_m1 < THRESH_M1
approved_xgb_test = y_prob_test_xgb < THRESH_XGB

lo, hi = ci(boot['profit_m0b'])
print(f'M0b (approve all): profit = $ {profit_m0b_test:,.2f} | 95% CI [$ {lo:,.2f}, $ {hi:,.2f}]')
print(f'  % approved: 100.00% | default among approved: {default_rate_test * 100:.4f}%')
print()

auc_m1 = roc_auc_score(y_test, y_prob_test_m1)
auc_pr_m1 = average_precision_score(y_test, y_prob_test_m1)
brier_m1 = brier_score_loss(y_test, y_prob_test_m1)
profit_m1_test = profit_at_threshold(y_test, y_prob_test_m1, THRESH_M1, interest_test, loss_test)
lo_auc, hi_auc = ci(boot['auc_m1'])
lo_brier, hi_brier = ci(boot['brier_m1'])
lo_profit, hi_profit = ci(boot['profit_m1'])
print(f'M1 (threshold={THRESH_M1}):')
print(f'  AUC-ROC = {auc_m1:.4f} | 95% CI [{lo_auc:.4f}, {hi_auc:.4f}]')
print(f'  AUC-PR = {auc_pr_m1:.4f}')
print(f'  Brier = {brier_m1:.4f} | 95% CI [{lo_brier:.4f}, {hi_brier:.4f}]')
print(f'  Profit = $ {profit_m1_test:,.2f} | 95% CI [$ {lo_profit:,.2f}, $ {hi_profit:,.2f}]')
print(f'  % approved = {approved_m1_test.mean() * 100:.2f}% | default among approved = {y_test[approved_m1_test].mean() * 100:.4f}%')
print()

auc_xgb = roc_auc_score(y_test, y_prob_test_xgb)
auc_pr_xgb = average_precision_score(y_test, y_prob_test_xgb)
brier_xgb = brier_score_loss(y_test, y_prob_test_xgb)
profit_xgb_test = profit_at_threshold(y_test, y_prob_test_xgb, THRESH_XGB, interest_test, loss_test)
lo_auc_x, hi_auc_x = ci(boot['auc_xgb'])
lo_brier_x, hi_brier_x = ci(boot['brier_xgb'])
lo_profit_x, hi_profit_x = ci(boot['profit_xgb'])
print(f'XGB_walkforward (threshold={THRESH_XGB}):')
print(f'  AUC-ROC = {auc_xgb:.4f} | 95% CI [{lo_auc_x:.4f}, {hi_auc_x:.4f}]')
print(f'  AUC-PR = {auc_pr_xgb:.4f}')
print(f'  Brier = {brier_xgb:.4f} | 95% CI [{lo_brier_x:.4f}, {hi_brier_x:.4f}]')
print(f'  Profit = $ {profit_xgb_test:,.2f} | 95% CI [$ {lo_profit_x:,.2f}, $ {hi_profit_x:,.2f}]')
print(f'  % approved = {approved_xgb_test.mean() * 100:.2f}% | default among approved = {y_test[approved_xgb_test].mean() * 100:.4f}%')


M0b (approve all): profit = $ 233,202,813.06 | 95% CI [$ 228,626,556.25, $ 237,526,093.71]
  % approved: 100.00% | default among approved: 14.8836%

M1 (threshold=0.38):
  AUC-ROC = 0.6847 | 95% CI [0.6820, 0.6874]
  AUC-PR = 0.2617
  Brier = 0.1210 | 95% CI [0.1200, 0.1220]
  Profit = $ 235,936,408.63 | 95% CI [$ 231,368,999.00, $ 240,458,423.72]
  % approved = 98.88% | default among approved = 14.6184%



XGB_walkforward (threshold=0.31):
  AUC-ROC = 0.6846 | 95% CI [0.6818, 0.6871]
  AUC-PR = 0.2636
  Brier = 0.1205 | 95% CI [0.1195, 0.1214]
  Profit = $ 242,230,710.89 | 95% CI [$ 237,888,392.29, $ 246,720,384.43]
  % approved = 96.24% | default among approved = 14.0492%


In [11]:
test_results_df = pd.DataFrame([
    {'model': 'M0b', 'auc_roc': np.nan, 'brier': np.nan, 'threshold': np.nan,
     'profit': profit_m0b_test, 'profit_ci_low': ci(boot['profit_m0b'])[0], 'profit_ci_high': ci(boot['profit_m0b'])[1],
     'pct_approved': 100.0, 'default_rate_approved': default_rate_test * 100},
    {'model': 'M1', 'auc_roc': auc_m1, 'brier': brier_m1, 'threshold': THRESH_M1,
     'profit': profit_m1_test, 'profit_ci_low': lo_profit, 'profit_ci_high': hi_profit,
     'pct_approved': approved_m1_test.mean() * 100, 'default_rate_approved': y_test[approved_m1_test].mean() * 100},
    {'model': 'XGB_walkforward', 'auc_roc': auc_xgb, 'brier': brier_xgb, 'threshold': THRESH_XGB,
     'profit': profit_xgb_test, 'profit_ci_low': lo_profit_x, 'profit_ci_high': hi_profit_x,
     'pct_approved': approved_xgb_test.mean() * 100, 'default_rate_approved': y_test[approved_xgb_test].mean() * 100},
]).set_index('model')
test_results_df


,auc_roc,brier,threshold,profit,profit_ci_low,profit_ci_high,pct_approved,default_rate_approved
model,,,,,,,,
M0b,NaN,NaN,NaN,2.332028e+08,2.286266e+08,2.375261e+08,100.000000,14.883640
M1,0.684718,0.121035,0.38,2.359364e+08,2.313690e+08,2.404584e+08,98.877954,14.618367
XGB_walkforward,0.684563,0.120469,0.31,2.422307e+08,2.378884e+08,2.467204e+08,96.236036,14.049231


## Section 3 - Bootstrap on the test set: differences

In [12]:
diff_m1_m0b = boot['profit_m1'] - boot['profit_m0b']
diff_xgb_m0b = boot['profit_xgb'] - boot['profit_m0b']
diff_m1_xgb = boot['profit_m1'] - boot['profit_xgb']

for name, d in [('M1 vs M0b', diff_m1_m0b), ('XGB_walkforward vs M0b', diff_xgb_m0b)]:
    lo, hi = ci(d)
    print(f'{name}: mean = $ {d.mean():,.2f} | 95% CI [$ {lo:,.2f}, $ {hi:,.2f}] | % positive = {(d > 0).mean() * 100:.2f}%')

print()
lo, hi = ci(diff_m1_xgb)
crosses_zero = lo < 0 < hi
print(f'M1 vs XGB_walkforward (paired): mean = $ {diff_m1_xgb.mean():,.2f} | 95% CI [$ {lo:,.2f}, $ {hi:,.2f}]')
print(f'  Crosses zero? {crosses_zero}')
print(f'  % of resamples where M1 won: {(diff_m1_xgb > 0).mean() * 100:.2f}%')
print(f'  % of resamples where XGB_walkforward won: {(diff_m1_xgb < 0).mean() * 100:.2f}%')


M1 vs M0b: mean = $ 2,750,813.30 | 95% CI [$ 2,037,165.24, $ 3,542,444.80] | % positive = 100.00%
XGB_walkforward vs M0b: mean = $ 9,068,587.76 | 95% CI [$ 7,627,319.06, $ 10,664,983.20] | % positive = 100.00%

M1 vs XGB_walkforward (paired): mean = $ -6,317,774.45 | 95% CI [$ -7,706,571.79, $ -4,931,521.12]
  Crosses zero? False
  % of resamples where M1 won: 0.00%
  % of resamples where XGB_walkforward won: 100.00%


**Mandatory reading rule:** if the paired M1 vs XGB_walkforward CI crosses zero, the two
models are statistically indistinguishable in test profit - no winner is declared
on profit, and the choice between them comes down to explainability/robustness, not to
a performance difference that the data supports.

## Section 4 - Validation -> Test: how much the gain shrank

Gain of each model over M0b, comparing the number already known from validation (recomputed
here with the SAME frozen model, only to ensure consistency - there was no re-tuning) with
the test number.

In [13]:
profit_m0b_val = interest_val[y_val == 0].sum() - loss_val[y_val == 1].sum()
gain_val_m1 = profit_at_threshold(y_val, y_prob_val_m1, THRESH_M1, interest_val, loss_val) - profit_m0b_val
gain_val_xgb = profit_at_threshold(y_val, y_prob_val_xgb, THRESH_XGB, interest_val, loss_val) - profit_m0b_val

gain_test_m1 = profit_m1_test - profit_m0b_test
gain_test_xgb = profit_xgb_test - profit_m0b_test

shrink_rows = []
for name, gv, gt in [('M1', gain_val_m1, gain_test_m1), ('XGB_walkforward', gain_val_xgb, gain_test_xgb)]:
    shrinkage = gv - gt
    pct = (shrinkage / gv * 100) if gv != 0 else np.nan
    shrink_rows.append({'model': name, 'validation_gain_$': gv, 'test_gain_$': gt,
                         'gain_shrinkage_$': shrinkage, 'gain_shrinkage_%': pct})
shrink_df = pd.DataFrame(shrink_rows).set_index('model')
shrink_df


,validation_gain_$,test_gain_$,gain_shrinkage_$,gain_shrinkage_%
model,,,,
M1,874222.01,2733595.57,-1859373.56,-212.688944
XGB_walkforward,1865541.47,9027897.83,-7162356.36,-383.929088


## Section 5 - Breakdown of the test result (highest-profit model)

In [14]:
best_model = 'M1' if profit_m1_test >= profit_xgb_test else 'XGB_walkforward'
print(f'Model with the highest test profit: {best_model}')

if best_model == 'M1':
    y_prob_best = y_prob_test_m1
    thresh_best = THRESH_M1
    profit_best = profit_m1_test
else:
    y_prob_best = y_prob_test_xgb
    thresh_best = THRESH_XGB
    profit_best = profit_xgb_test

decomp_result = decompose(y_prob_best, thresh_best, y_test, interest_test, loss_test, loan_amnt_test)
for k, v in decomp_result.items():
    if isinstance(v, float):
        print(f'{k}: $ {v:,.2f}' if 'gain' in k or 'loss' in k or 'interest' in k or 'amount' in k else f'{k}: {v}')
    else:
        print(f'{k}: {v}')

net_gain_check = decomp_result['net_gain_$']
expected_gain = profit_best - profit_m0b_test
print()
print(f'Net gain (decomposition): $ {net_gain_check:,.2f}')
print(f'Model profit - M0b profit (Section 2): $ {expected_gain:,.2f}')
print(f'Match (difference < $1)? {abs(net_gain_check - expected_gain) < 1}')


Model with the highest test profit: XGB_walkforward
threshold: 0.31
n_rejected: 10644
total_rejected_amount: $ 136,302,375.00
n_rejected_charged_off: 3855
n_rejected_fully_paid: 6789
avoided_loss_$: $ 32,151,026.71
lost_interest_$: $ 23,123,128.88
net_gain_$: $ 9,027,897.83

Net gain (decomposition): $ 9,027,897.83
Model profit - M0b profit (Section 2): $ 9,027,897.83
Match (difference < $1)? True


## Section 6 - Transfer analysis (60 months)

Both models were trained on 36-month loans (train.parquet). Here they are applied, without
any re-tuning, to `transfer_60m.parquet`.

In [15]:
X_transfer = prepare_X(transfer_60m, FEATURE_SET, CATEGORICAL_COLS)
cols_transfer = X_transfer.columns.tolist()
if cols_transfer != cols_train:
    only_train_t = set(cols_train) - set(cols_transfer)
    only_transfer = set(cols_transfer) - set(cols_train)
    print('Warning: transfer_60m columns differ from train.')
    print('Only in train:', only_train_t, '| Only in transfer_60m:', only_transfer)
    X_transfer = X_transfer.reindex(columns=X_train.columns, fill_value=0)
else:
    print('OK: transfer_60m columns identical to train.')

X_transfer_scaled = pd.DataFrame(scaler.transform(X_transfer), columns=X_transfer.columns, index=X_transfer.index)
y_transfer = transfer_60m['target'].values
interest_transfer, loss_transfer = compute_financials(transfer_60m)
interest_transfer = interest_transfer.values
loss_transfer = loss_transfer.values

default_rate_transfer = transfer_60m['target'].mean()
print(f'N transfer_60m: {len(transfer_60m)} | default rate: {default_rate_transfer * 100:.2f}%')
print(f'Default rate in test (36m): {default_rate_test * 100:.2f}%')


OK: transfer_60m columns identical to train.


N transfer_60m: 54969 | default rate: 25.16%
Default rate in test (36m): 14.88%


In [16]:
y_prob_transfer_m1 = m1_model.predict_proba(X_transfer_scaled)[:, 1]
y_prob_transfer_xgb = xgb_wf.predict_proba(X_transfer)[:, 1]

profit_m0b_transfer = interest_transfer[y_transfer == 0].sum() - loss_transfer[y_transfer == 1].sum()

auc_transfer_m1 = roc_auc_score(y_transfer, y_prob_transfer_m1)
profit_transfer_m1 = profit_at_threshold(y_transfer, y_prob_transfer_m1, THRESH_M1, interest_transfer, loss_transfer)
gain_transfer_m1 = profit_transfer_m1 - profit_m0b_transfer

auc_transfer_xgb = roc_auc_score(y_transfer, y_prob_transfer_xgb)
profit_transfer_xgb = profit_at_threshold(y_transfer, y_prob_transfer_xgb, THRESH_XGB, interest_transfer, loss_transfer)
gain_transfer_xgb = profit_transfer_xgb - profit_m0b_transfer

transfer_rows = []
for name, auc_36, auc_60, gain_36, gain_60 in [
    ('M1', auc_m1, auc_transfer_m1, gain_test_m1, gain_transfer_m1),
    ('XGB_walkforward', auc_xgb, auc_transfer_xgb, gain_test_xgb, gain_transfer_xgb),
]:
    transfer_rows.append({
        'model': name, 'auc_roc_36m': auc_36, 'auc_roc_60m': auc_60, 'delta_auc': auc_60 - auc_36,
        'gain_vs_m0b_36m_$': gain_36, 'gain_vs_m0b_60m_$': gain_60,
        'delta_gain_%': (gain_60 - gain_36) / gain_36 * 100 if gain_36 != 0 else np.nan,
    })
transfer_df = pd.DataFrame(transfer_rows).set_index('model')
transfer_df


,auc_roc_36m,auc_roc_60m,delta_auc,gain_vs_m0b_36m_$,gain_vs_m0b_60m_$,delta_gain_%
model,,,,,,
M1,0.684718,0.61198,-0.072738,2733595.57,-20472.60,-100.748926
XGB_walkforward,0.684563,0.64332,-0.041243,9027897.83,1066477.36,-88.186869


**Reading:** mild degradation of AUC/gain between 36m and 60m suggests that the two terms
share the same underlying risk behavior (a single scorecard would be enough).
Severe degradation suggests structurally different populations (a finding about the dataset,
not a model failure) - the base default rate is already visibly higher in 60m than
in 36m (see the cell above), which by itself points to populations with a different risk
composition.

## Section 7 - Reading notes

**This is the final result of the project. There is no re-tuning after this note.**

**(a) Which model has the highest test profit:** XGB_walkforward, with $242,230,710.89
versus $235,936,408.63 for M1 - an advantage of $6,317,774.45 at each model's already
frozen threshold (0.31 and 0.38, respectively). The two test AUC-ROCs are
practically identical (0.6846 vs 0.6847) - XGB_walkforward's advantage lies almost
entirely in how profit accumulates near the cut-off point, not in the
ability to rank risk in general.

**(b) Is the difference statistically distinguishable:** YES, clearly. The paired 95% CI
(M1 - XGB_walkforward) is [-$7,706,571.79, -$4,931,521.12] - it does not cross zero,
and XGB_walkforward won in 100% of the 1,000 resamples. This contrasts with the more
cautious reading that was needed on validation (notebooks 10/11, where comparisons between
XGBoost configurations had narrower margins and were subject to rerun/data-order
noise). Here, with the larger test population (282,787 rows) and a
profit advantage of almost 3% of total profit, the distinction is robust - **XGB_walkforward
is declared the winner on profit in this evaluation.**

**(c) How much the gain shrank, validation -> test:** it did NOT shrink - it grew, and
sharply, for both models:
- M1: from $874,222.01 (validation) to $2,733,595.57 (test) - growth of +213%.
- XGB_walkforward: from $1,865,541.47 (validation) to $9,027,897.83 (test) -
  growth of +384%.

This is the OPPOSITE of what "overfitting of the selection process" would predict (the usual
expectation is that the gain measured on validation shrinks on test, not that it grows). Part of the
growth is mechanical - the test set has ~74% more rows than validation (282,787 vs
162,570), so the absolute dollar gain tends to scale with portfolio size.
But even in RELATIVE terms (gain / M0b profit), the advantage grew: XGB_walkforward
went from 0.98% of M0b profit on validation to 3.87% on test. That is
good news - the process of following validation up to this point (walk-forward, calibration,
quantifying rerun noise) does not seem to have produced a model that only
worked on validation - but it is a SINGLE test, with no repetition: from
this result alone there is no way to know whether the growth in gain is a pattern that would repeat in new vintages or
a favorable quirk of the 2015 vintage. One should not extrapolate that the gain will
keep growing.

**(d) What the transfer to 60 months revealed:** SEVERE degradation in both models,
not mild. The base default rate almost doubles (25.16% in 60m versus 14.88% in 36m on the
test set). AUC-ROC drops in both (M1: 0.6847 -> 0.6120, -7.3 percentage points;
XGB_walkforward: 0.6846 -> 0.6433, -4.1 percentage points), and the gain over M0b almost
disappears or turns negative:
- M1: from +$2,733,595.57 (36m) to **-$20,472.60 (60m)** - worse than approving everyone.
- XGB_walkforward: from +$9,027,897.83 (36m) to +$1,066,477.36 (60m) - it loses 88% of the
  gain, but is still positive.

By the reading rule stated earlier in this notebook, this is evidence of structurally
different populations between 36 and 60 months - not the same risk behavior. A
scorecard trained on 36-month loans should not be used, without recalibration or
a dedicated model, to decide on 60-month loans: that is a finding about the
structure of the product/dataset, not a failure of the models. It is worth noting that XGB_walkforward
degrades LESS than M1 on both axes (AUC and profit) - its decision boundary generalizes
slightly better to the higher-risk population, although still severely.

**(e) Dataset selection bias - mandatory reading:** every profit number reported
in this notebook, from the start to the end of the project, measures performance on the population of
loans that LendingClub ALREADY APPROVED and originated. The model never saw, and cannot be
evaluated on, the applicants that LC rejected before even granting the loan.
This means the profit numbers here estimate how much value a SECOND FILTER, applied
on top of the portfolio LC has already decided to approve, would add - NOT an estimate of
how much profit a credit-granting system built from scratch, deciding on ALL
applicants (approved and rejected by LC), would generate. Extrapolating these numbers to
a credit policy that decides on the general applicant population would overestimate
the result - the real applicant population presumably includes higher-risk
cases that LC already filtered out and that none of the models here has ever seen.

**Final result, no re-tuning:** XGB_walkforward (max_depth=8, learning_rate=0.03,
n_estimators=600, min_child_weight=10, subsample=0.8, colsample_bytree=0.6, threshold
0.31) is the highest-profit model on the 2015 test, with a statistically
distinguishable advantage over M1. The transfer to 60 months shows that this result does NOT
generalize to longer-term loans without additional work. The measured profit
reflects only the population already approved by LendingClub.